protocol/mqtt_paho/v2/protocol.go

```diff
@@ -39,35 +41,47 @@ var (
 	_ protocol.Closer   = (*Protocol)(nil)
 )
 
-func New(ctx context.Context, clientConfig *paho.ClientConfig, connConfig *paho.Connect,
-	publishOpt *paho.Publish, subscribeOpt *paho.Subscribe,
-) (*Protocol, error) {
-	client := paho.NewClient(*clientConfig)
-	ca, err := client.Connect(ctx, connConfig)
-	if err != nil {
-		return nil, err
+func New(config *paho.ClientConfig, opts ...Option) (*Protocol, error) {
+	p := &Protocol{
+		config: config,
+		// default connect option
+		connOption: &paho.Connect{
+			KeepAlive:  30,
+			CleanStart: true,
+		},
+		incoming:    make(chan *paho.Publish),
+		openerMutex: sync.Mutex{},
+		clientMutex: sync.Mutex{},
+		closeChan:   make(chan struct{}),
 	}
-	if ca.ReasonCode != 0 {
-		return nil, fmt.Errorf("failed to connect to %s : %d - %s", client.Conn.RemoteAddr(), ca.ReasonCode,
-			ca.Properties.ReasonString)
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
 	}
+	return p, nil
+}
 
-	return &Protocol{
-		client:          client,
-		connConfig:      connConfig,
-		publishOption:   publishOpt,
-		subscribeOption: subscribeOpt,
-		incoming:        make(chan *paho.Publish),
-		openerMutex:     sync.Mutex{},
-		closeChan:       make(chan struct{}),
-	}, nil
+func (p *Protocol) applyOptions(opts ...Option) error {
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (t *Protocol) Send(ctx context.Context, m binding.Message, transformers ...binding.Transformer) error {
-	var err error
+func (p *Protocol) Send(ctx context.Context, m binding.Message, transformers ...binding.Transformer) error {
+	if p.publishOption == nil {
+		return fmt.Errorf("the paho.Publish option should not be nil")
+	}
+
+	err := p.initClient(ctx)
```

Comment:
question: this is called in `Send` and `OpenInbound` which might be confusing/inconsistent across usage. Should `initClient` be part of `New()`?

Reply:
That's okay. Just if we put `initClient` on `New()`, then the `ctx` must be added as the parameter of the `New()` constructor.

Reply:
I don't think this is an issue and makes the protocol even easier to use/harder to misuse